crates/wasi-keyvalue/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+
+    async fn increment(&mut self, key: String, delta: u64) -> Result<u64, Error>;
+
+    async fn get_many(
+        &mut self,
+        keys: Vec<String>,
+    ) -> Result<Vec<Option<(String, Vec<u8>)>>, Error>;
+
+    async fn set_many(&mut self, key_values: Vec<(String, Vec<u8>)>) -> Result<(), Error>;
+
+    async fn delete_many(&mut self, keys: Vec<String>) -> Result<(), Error>;
+}
+
+/// Builder-style structure used to create a [`WasiKeyValueCtx`].
+#[derive(Default)]
+pub struct WasiKeyValueCtxBuilder {
+    #[cfg(feature = "redis")]
+    redis_connection_timeout: Option<std::time::Duration>,
+    #[cfg(feature = "redis")]
+    redis_response_timeout: Option<std::time::Duration>,
+}
+
+impl WasiKeyValueCtxBuilder {
+    /// Creates a builder for a new context with default parameters set.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Sets the connection timeout parameter for the Redis provider.
+    #[cfg(feature = "redis")]
+    pub fn redis_connection_timeout(mut self, t: std::time::Duration) -> Self {
+        self.redis_connection_timeout = Some(t);
+        self
+    }
+
+    /// Sets the response timeout parameter for the Redis provider.
+    #[cfg(feature = "redis")]
+    pub fn redis_response_timeout(mut self, t: std::time::Duration) -> Self {
+        self.redis_response_timeout = Some(t);
+        self
+    }
+
+    /// Uses the configured context so far to construct the final [`WasiKeyValueCtx`].
+    pub fn build(self) -> WasiKeyValueCtx {
+        WasiKeyValueCtx {
+            #[cfg(feature = "redis")]
+            redis_connection_timeout: self.redis_connection_timeout,
+            #[cfg(feature = "redis")]
+            redis_response_timeout: self.redis_response_timeout,
+        }
+    }
+}
+
+/// Capture the state necessary for use in the `wasi-keyvalue` API implementation.
+pub struct WasiKeyValueCtx {
+    #[cfg(feature = "redis")]
+    redis_connection_timeout: Option<std::time::Duration>,
+    #[cfg(feature = "redis")]
+    redis_response_timeout: Option<std::time::Duration>,
+}
+
+impl WasiKeyValueCtx {
+    /// Convenience function for calling [`WasiKeyValueCtxBuilder::new`].
+    pub fn builder() -> WasiKeyValueCtxBuilder {
+        WasiKeyValueCtxBuilder::new()
+    }
+}
+
+/// A wrapper capturing the needed internal `wasi-keyvalue` state.
+pub struct WasiKeyValue<'a> {
+    ctx: &'a WasiKeyValueCtx,
+    table: &'a mut ResourceTable,
+}
+
+impl<'a> WasiKeyValue<'a> {
+    /// Create a new view into the `wasi-keyvalue` state.
+    pub fn new(ctx: &'a WasiKeyValueCtx, table: &'a mut ResourceTable) -> Self {
+        Self { ctx, table }
+    }
+}
+
+#[async_trait]
+impl keyvalue::store::Host for WasiKeyValue<'_> {
+    async fn open(&mut self, identifier: String) -> Result<Resource<Bucket>, Error> {
+        if identifier == "" {
+            return Ok(self.table.push(Bucket {
+                inner: Box::new(provider::inmemory::InMemory::default()),
```

Comment:
Since `::default()` is used here it looks like there's no way to configure this in-memory configuration?

Reply:
Yes, the in-memory implementation is very simple. Currently, it doesn't have any configurable features, so configuration is not supported here.

Reply:
I'm at least personally a bit hesitant to include this when it doesn't have any support/tests/etc. Would it be possible to hook this up to the configuration of `WasiKeyValueCtx`? Or somehow expose the ability to configure this? Otherwise I think it'd be better to remove the internals of `InMemory` and make it an empty struct which is a noop implementation of the trait.

Reply:
I added a configuration option to the In-Memory provider that allows presetting some data. I believe this provider is valuable, especially in testing scenarios, where users can use this provider to mock some data without relying on a real database. PTAL if this is reasonable.